impl/src/main/kotlin/nl/nuts/consent/bridge/listener/StateFileStorage.kt

```diff
@@ -0,0 +1,155 @@
+
+    /**
+     * gets correct stateFileStorage or creates a new one if needed
+     */
+    private fun getOrCreateStateFileStorage(eventType: String) : StateFileStorage {
+        synchronized(this) {
+            var stateFileStorage = stateFileStorages[eventType]
+            if (stateFileStorage == null) {
+                stateFileStorage = StateFileStorage(eventTypeFileLocation(eventType))
+                stateFileStorages[eventType] = stateFileStorage
+            }
+            return stateFileStorage
+        }
+    }
+
+    /**
+     * returns correct File handler for given eventType, making sure it exists and is accessible
+     * must be called from synchronized context
+     */
+    private fun eventTypeFileLocation(eventType: String): File {
+        val dir = File(eventMetaProperties.location)
+        if (!dir.exists()) {
+            dir.mkdirs() // throws SecurityException is access denied
+        }
+
+        val file = dir.resolve("$eventType.$FILE_EXTENSION")
+        if (!file.exists()) {
+            file.createNewFile() // throws IOException if denied/wrong
+        }
+
+        return file
+    }
+}
+
+/**
+ * Handles writing and reading a timestamp to/from a given file with locking
+ */
+class StateFileStorage(val file: File) {
+
+    /**
+     * Writes latest timestamp to file if larger then existing timestamp
+     */
+    fun writeTimestamp(timestamp: Long) {
+        withFileLock { stream ->
+            val currentTimestamp = readTimestamp(stream)
+            if (timestamp > currentTimestamp) {
+                stream.seek(0)
+                // overwrites everything
+                stream.writeBytes("$timestamp")
+            }
+        }
+    }
+
+    /**
+     * Reads latest timestamp from file
+     */
+    fun readTimestamp(): Long {
+        var currentTimestamp = 0L
+
+        withFileLock { stream ->
+            currentTimestamp = readTimestamp(stream)
+        }
+
+        return currentTimestamp
+    }
+
+    private fun readTimestamp(stream: RandomAccessFile): Long {
+        var currentTimestamp = 0L
+
+        stream.seek(0)
+        val b = ByteArray(13)
+        try {
+            val read = stream.read(b)
+            if (read >= 13) {
+                currentTimestamp = String(b).trim().toLong()
+            }
+        } catch (e: EOFException) {
+            // use 0L as value
+        } catch (e: NumberFormatException) {
+            // use 0L as value
+        }
+        return currentTimestamp
+    }
+
+    private fun withFileLock(block: (stream: RandomAccessFile) -> Unit) {
+        synchronized(this) {
+            val stream = RandomAccessFile(file, "rw")
+            val channel: FileChannel = stream.channel
+
+            // IllegalState or IOException can be thrown
+            var lock: FileLock? = null
+            try {
+                lock = channel.lock()
+                block(stream)
+            } finally {
+                lock?.release()
+                stream.close()
```

Comment:
instead of manually closing, you could use Kotlin's 'use' block functioN:

```
RandomAccessFile(file, "rw").use { stream ->
    stream.channel.use { channel ->
        // IllegalState or IOException can be thrown
        var lock: FileLock? = null
        try {
            lock = channel.lock()
            block(stream)
        } finally {
            lock?.release()
        }
    }
}
```